File: android/app/src/main/java/com/neuro/ac/sdk/DeepLinkRouter.kt
package com.neuro.ac.sdk

import android.content.Intent
import android.net.Uri

sealed class DeepLinkDestination {
    data class Subscription(val tier: String?) : DeepLinkDestination()
    data class ReferralRegister(val inviterId: String) : DeepLinkDestination()
    data object Credits : DeepLinkDestination()
    data object Home : DeepLinkDestination()
}

class DeepLinkRouter {
    companion object {
        private const val CUSTOM_SCHEME = "neuro"
        private const val UNIVERSAL_HOST = "ac.neuro.ai"
    }

    fun route(intent: Intent): DeepLinkDestination? {
        val data = intent.data ?: return null
        return route(data)
    }

    fun route(uri: Uri): DeepLinkDestination? {
        return when (uri.scheme) {
            CUSTOM_SCHEME -> parseCustomScheme(uri)
            "https" -> if (uri.host == UNIVERSAL_HOST) parseUniversalLink(uri) else null
            "http" -> if (uri.host == UNIVERSAL_HOST) parseUniversalLink(uri) else null
            else -> null
        }
    }

    private fun parseCustomScheme(uri: Uri): DeepLinkDestination {
        val host = uri.host ?: return DeepLinkDestination.Home
        val params = extractQueryParams(uri)

        return when (host) {
            "subscribe" -> DeepLinkDestination.Subscription(tier = params["tier"])
            "register" -> {
                val inviterId = params["inviter_id"]
                if (inviterId != null) DeepLinkDestination.ReferralRegister(inviterId)
                else DeepLinkDestination.Home
            }
            "referral" -> {
                val inviterId = params["inviter_id"]
                if (inviterId != null) DeepLinkDestination.ReferralRegister(inviterId)
                else DeepLinkDestination.Home
            }
            "credits" -> DeepLinkDestination.Credits
            else -> DeepLinkDestination.Home
        }
    }

    private fun parseUniversalLink(uri: Uri): DeepLinkDestination {
        val segments = uri.pathSegments
        if (segments.isEmpty()) return DeepLinkDestination.Home
        val params = extractQueryParams(uri)

        return when (segments[0]) {
            "subscribe" -> {
                val tier = params["tier"] ?: segments.getOrNull(1)
                DeepLinkDestination.Subscription(tier = tier)
            }
            "register" -> {
                val inviterId = params["inviter_id"]
                if (inviterId != null) DeepLinkDestination.ReferralRegister(inviterId)
                else DeepLinkDestination.Home
            }
            "referral" -> {
                val inviterId = params["inviter_id"]
                if (inviterId != null) DeepLinkDestination.ReferralRegister(inviterId)
                else DeepLinkDestination.Home
            }
            "credits" -> DeepLinkDestination.Credits
            else -> DeepLinkDestination.Home
        }
    }

    private fun extractQueryParams(uri: Uri): Map<String, String> {
        val params = mutableMapOf<String, String>()
        for (paramName in uri.queryParameterNames) {
            uri.getQueryParameter(paramName)?.let { value ->
                params[paramName] = value
            }
        }
        return params
    }
}

File: ios/NeuroAC/Sources/Routing/DeepLinkRouter.swift
import Foundation

enum DeepLinkDestination {
    case subscription(tier: String?)
    case referralRegister(inviterId: String)
    case credits
    case home
}

class DeepLinkRouter {
    static let shared = DeepLinkRouter()

    private let supportedHosts = ["subscribe", "register", "credits", "referral"]
    private let universalLinkHost = "ac.neuro.ai"

    func route(url: URL) -> DeepLinkDestination? {
        if url.scheme == "neuro" {
            return parseCustomScheme(url)
        } else if url.scheme == "https", url.host == universalLinkHost {
            return parseUniversalLink(url)
        }
        return nil
    }

    private func parseCustomScheme(_ url: URL) -> DeepLinkDestination? {
        guard let host = url.host else { return nil }
        let params = extractQueryParams(url)

        switch host {
        case "subscribe":
            let tier = params["tier"]
            return .subscription(tier: tier)
        case "register":
            if let inviterId = params["inviter_id"] {
                return .referralRegister(inviterId: inviterId)
            }
            return .home
        case "referral":
            if let inviterId = params["inviter_id"] {
                return .referralRegister(inviterId: inviterId)
            }
            return .home
        case "credits":
            return .credits
        default:
            return .home
        }
    }

    private func parseUniversalLink(_ url: URL) -> DeepLinkDestination? {
        let pathComponents = url.pathComponents.filter { $0 != "/" }

        guard pathComponents.count >= 1 else { return .home }
        let params = extractQueryParams(url)

        switch pathComponents[0] {
        case "subscribe":
            let tier = params["tier"] ?? (pathComponents.count > 1 ? pathComponents[1] : nil)
            return .subscription(tier: tier)
        case "register":
            if let inviterId = params["inviter_id"] {
                return .referralRegister(inviterId: inviterId)
            }
            return .home
        case "referral":
            if let inviterId = params["inviter_id"] {
                return .referralRegister(inviterId: inviterId)
            }
            return .home
        case "credits":
            return .credits
        default:
            return .home
        }
    }

    private func extractQueryParams(_ url: URL) -> [String: String] {
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              let queryItems = components.queryItems else { return [:] }
        return queryItems.reduce(into: [String: String]()) { result, item in
            if let value = item.value {
                result[item.name] = value
            }
        }
    }
}
